### ISSUE TITLE
Preview diff not shown for Python dynamic resources providers

### ISSUE THREAD

**Issue by felix11h** [pulumi/pulumi#12738]
### What happened?

For dynamic resource providers written in Python, I want to see a preview of changes to my resources when calling `pulumi preview --diff` or in the details of `pulumi up`. This has worked in previous versions of Pulumi (for example 3.51.0), but does not work in newer versions (for example 3.64.0).

### Expected Behavior

`pulumi preview --diff` shoes detailed changes like in previous versions of Pulumi.

### Steps to reproduce

I am using the following code to reproduce the issue:

```python
from typing import Any
from uuid import uuid4

from pulumi import Output
from pulumi.dynamic import CreateResult, DiffResult, Resource, ResourceProvider, UpdateResult


class TestResourceProvider(ResourceProvider):

    def create(self, props: Any) -> CreateResult:
        return CreateResult(id_=uuid4().hex, outs=props)

    def diff(self, _id: str, _olds: Any, _news: Any) -> DiffResult:
        return DiffResult(changes=_olds["content"] != _news["content"])

    def update(self, _id: str, _olds: Any, _news: Any) -> UpdateResult:
        return UpdateResult(outs={**_news})


class LocalFile(Resource):
    content: Output[str]

    def __init__(self, resource_name, content, opts=None):
        super().__init__(
            TestResourceProvider(),
            name=resource_name,
            props={"content": content},
            opts=opts,
        )


# changes in content correctly previewed in 3.51.0 but not 3.64.0
LocalFile(resource_name="test", content="hello world")
```

First initialize a new stack, call `pulumi up` once, and change the content string above to "hello world test". 

When using Pulumi 3.51.0, calling `pulumi up` and selecting details gives:

![Screenshot 2023-04-25 at 10 21 25](https://user-images.githubusercontent.com/610633/234235328-cebb5c2e-a324-4134-9c96-266d6efe72cd.png)

Using Pulumi 3.64.0, however, the changes in the details are not shown:

![Screenshot 2023-04-25 at 10 23 27](https://user-images.githubusercontent.com/610633/234235519-05f2e54b-2e65-4007-b802-cae01ac597cf.png)

The same is true for `pulumi preview --diff`, which works correctly in older versions but not in newer versions.

### Output of `pulumi about`

For Pulumi 3.64.0:

```
CLI          
Version      3.64.0
Go Version   go1.20.3
Go Compiler  gc

Plugins
NAME    VERSION
python  unknown

Host     
OS       darwin
Version  12.6
Arch     x86_64

This project is written in python: executable='/Users/felixhoffmann/Library/Caches/pypoetry/virtualenvs/pulumi-demos-nTYqTP6e-py3.9/bin/python3' version='3.9.1
'

Current Stack: test-03

TYPE                            URN
pulumi:pulumi:Stack             urn:pulumi:test-03::missing-preview::pulumi:pulumi:Stack::missing-preview-test-03
pulumi:providers:pulumi-python  urn:pulumi:test-03::missing-preview::pulumi:providers:pulumi-python::default
pulumi-python:dynamic:Resource  urn:pulumi:test-03::missing-preview::pulumi-python:dynamic:Resource::test


Found no pending operations associated with test-03

Backend        
Name           felixhoffmann
URL            file://.backend
User           felixhoffmann
Organizations  

Dependencies:
NAME                 VERSION
flake8-coding        1.3.2
flake8-commas        2.0.0
flake8-import-order  0.18.1
flake8-quotes        3.2.0
nose                 1.3.7
pip                  20.2.2
pulumi               2.24.1
wheel                0.35.1

Pulumi locates its logs in /var/folders/hf/b7_fk0k54rnd53m8rlf2kz1w0000gp/T/ by default
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by amazing-bettyanne** [comment]
same in version 3.68.0

**Comment by marns93** [comment]
@abhinav Any plans to fix this? It is very annoying to not see the correct diff.

**Comment by felix11h** [comment]
Fixed in 3.74.0. Thank you, @Frassle! 🙏

**Comment by amazing-bettyanne** [comment]
Bravo!
